query-service-impl/src/main/java/org/hypertrace/core/query/service/ExecutionContext.java

```diff
@@ -35,14 +39,62 @@
   // while the selectedColumns
   // is a set of column names.
   private final LinkedHashSet<Expression> allSelections;
+  private final Optional<Double> timeSeriesPeriod;
+  private final double timeRangeDuration;
 
   public ExecutionContext(String tenantId, QueryRequest request) {
     this.tenantId = tenantId;
     this.selectedColumns = new LinkedHashSet<>();
     this.allSelections = new LinkedHashSet<>();
+    this.timeSeriesPeriod = setTimeSeriesPeriod(request);
+    this.timeRangeDuration = setTimeRangeDuration(request);
     analyze(request);
   }
 
+  private Optional<Double> setTimeSeriesPeriod(QueryRequest request) {
+    Optional<Double> period = Optional.empty();
+    if (request.getGroupByCount() > 0) {
+      for (Expression expression : request.getGroupByList()) {
+        if (expression.getValueCase() == ValueCase.FUNCTION
+            && expression.getFunction().getFunctionName().equals("dateTimeConvert")) {
+          String periodInIso =
```

Comment:
this isn't an iso string, this is a pinot format like `30:seconds` I believe. ISO durations are described here: https://en.wikipedia.org/wiki/ISO_8601#Durations (e.g. `P30S`)

Reply:
Yeah currently the request comes from gateway service so has period in the form : `30:SECONDS`. This has to be changed over there.

Reply:
The dateTimeConvert function takes other sets of args too as an example below. And for now, it is directly mapped to the pinot dateTimeConvert function argument format. So, I would suggest not making changes to this function for now. We can do this as a separate ticket. This will requires changes in the caller too. And, we anyway have to move this conversation into QueryService for dataTimeConvert instead of a gateway.
```
function {
    functionName: "dateTimeConvert"
    arguments {
      columnIdentifier {
        columnName: "BACKEND.startTime"
      }
    }
    arguments {
      literal {
        value {
          string: "1:MILLISECONDS:EPOCH"
        }
      }
    }
    arguments {
      literal {
        value {
          string: "1:MILLISECONDS:EPOCH"
        }
      }
    }
    arguments {
      literal {
        value {
          string: "30:SECONDS"
        }
      }
    }
  }
```

Reply:
I agree, we shouldn't touch dateTimeConvert's format (and it should eventually be moved down), but we're mixing things. We don't need to use the pinot format that dateTimeConvert uses in the two scenarios we're talking here:
1. AVG_RATE arguments - this is a new function from gateway's perspective, so it will need to write code to use it no matter what we do. Might as well make it use the standard string format.
2. Internally - we should be using neither string format, and using duration objects. My comment here was about the variable name, since the value being put into `String periodInIso` was something of the format `30:SECONDS` - not an ISO string.

Reply:
@kotharironak should I create an issue for this ?

Reply:
@sarthak77 we can create two tickets.
One for enhancing query service APIs (need to figure out the details though)
- for accepting time range at the top level
- aggregation API vs series API
Second, moving dateTimeConvert conversation from the gateway to the pinot request handler will be a separate issue.  This we can take care of when we separate out time-series API and aggregation API.